src/stream/src/executor/hash_agg.rs

```diff
@@ -274,53 +286,69 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
             .collect()
     }
 
-    async fn ensure_keys_in_cache(
+    /// Touch the [`AggGroup`]s for the given keys, which means move them from cache to the `dirty_groups` map.
+    /// If the [`AggGroup`] doesn't exist in the cache before, it will be created or recovered from state table.
+    async fn touch_agg_groups(
         this: &ExecutorInner<K, S>,
-        cache: &mut AggGroupCache<K, S>,
+        vars: &mut ExecutionVars<K, S>,
         keys: impl IntoIterator<Item = &K>,
-        stats: &mut ExecutionStats,
     ) -> StreamExecutorResult<()> {
         let group_key_types = &this.info.schema.data_types()[..this.group_key_indices.len()];
         let futs = keys
             .into_iter()
             .filter_map(|key| {
-                stats.total_lookup_count += 1;
-                if cache.contains(key) {
-                    None
-                } else {
-                    stats.lookup_miss_count += 1;
-                    Some(async {
-                        // Create `AggGroup` for the current group if not exists. This will
-                        // restore agg states from the intermediate state table.
-                        let agg_group = AggGroup::create(
-                            Some(GroupKey::new(
-                                key.deserialize(group_key_types)?,
-                                Some(this.group_key_table_pk_projection.clone()),
-                            )),
-                            &this.agg_calls,
-                            &this.agg_funcs,
-                            &this.storages,
-                            &this.intermediate_state_table,
-                            &this.input_pk_indices,
-                            this.row_count_index,
-                            this.extreme_cache_size,
-                            &this.input_schema,
-                        )
-                        .await?;
-                        Ok::<_, StreamExecutorError>((key.clone(), agg_group))
-                    })
+                vars.stats.total_lookup_count += 1;
+                if vars.dirty_groups.contains_key(key) {
+                    // already dirty
+                    return None;
+                }
+                match vars.agg_group_cache.get_mut(key) {
+                    Some(mut agg_group) => {
+                        let agg_group: &mut Option<_> = &mut agg_group;
```

Comment:
It seems to bypass the EstimatedLruCache's mut guard mechanism, maybe the lru cache's `kv_heap_size` should be updated?